### PR TITLE
IDE-4426 use ResourceChangeManager to dispatch all change events

### PR DESCRIPTION
**Pull request by Camork**

### CODE REVIEW

File: tools/plugins/com.liferay.ide.core/src/com/liferay/ide/core/LiferayCore.java [hunk 1]
```diff
@@ -252,23 +279,110 @@ public static void logWarning(Throwable t) {
...
+
+		_listenerServiceTracker = _createServiceTracker(context, LiferayListenerRegistry.class);
+
+		_listenerService = _registerService(
+			context, LiferayListenerRegistry.class, new DefaultListenerRegistry(), preferences);
+
+		_resourceChangeManager = LiferayResourceChangeManager.createAndRegister();
+	}
+
+	private void _unRegisterServices() {
```
Comment (jtydhr88) [inline tools/plugins/com.liferay.ide.core/src/com/liferay/ide/core/LiferayCore.java]: use _unregisterServices

File: tools/plugins/com.liferay.ide.core/src/com/liferay/ide/core/IResourceBundleProject.java [hunk 2]
```diff
@@ -21,7 +21,7 @@
 /**
  * @author Andy Wu
  */
-public interface IResourceBundleProject {
+public interface IResourceBundleProject extends ILiferayProject {
```
Comment (jtydhr88) [inline tools/plugins/com.liferay.ide.core/src/com/liferay/ide/core/IResourceBundleProject.java]: why do we need to change this class

Comment (Camork) [reply tools/plugins/com.liferay.ide.core/src/com/liferay/ide/core/IResourceBundleProject.java]: Greg did.

File: tools/plugins/com.liferay.ide.core/src/com/liferay/ide/core/LiferayCore.java (hunk 1 above)
Comment (Camork) [reply tools/plugins/com.liferay.ide.core/src/com/liferay/ide/core/LiferayCore.java]: sf not pass

File: tools/plugins/com.liferay.ide.core/src/com/liferay/ide/core/LiferayCore.java [hunk 3]
```diff
@@ -43,28 +54,42 @@
 
 	public static final String PLUGIN_ID = "com.liferay.ide.core";
 
-	public static <T> T create(Class<T> type, Object adaptable) {
+	public static synchronized <T extends ILiferayProject> T create(Class<T> type, Object adaptable) {
```
Comment (jtydhr88) [inline tools/plugins/com.liferay.ide.core/src/com/liferay/ide/core/LiferayCore.java]: why do we need to change here?

File: tools/plugins/com.liferay.ide.core/src/com/liferay/ide/core/LiferayCore.java [hunk 4]
```diff
@@ -252,23 +279,110 @@ public static void logWarning(Throwable t) {
...
 
 	@Override
 	public void stop(BundleContext context) throws Exception {
+		_unRegisterServices();
 		_plugin = null;
 
 		super.stop(context);
 	}
 
+	private ServiceTracker<?, ?> _createServiceTracker(BundleContext context, Class<?> clazz) {
```
Comment (jtydhr88) [inline tools/plugins/com.liferay.ide.core/src/com/liferay/ide/core/LiferayCore.java]: no need this method

File: tools/plugins/com.liferay.ide.core/src/com/liferay/ide/core/LiferayCore.java [hunk 5]
```diff
@@ -252,23 +279,110 @@ public static void logWarning(Throwable t) {
...
+		ServiceTracker<?, ?> serviceTracker = new ServiceTracker<>(context, clazz.getName(), null);
+
+		serviceTracker.open();
+
+		return serviceTracker;
+	}
+
+	private <T> ServiceRegistration<?> _registerService(
```
Comment (jtydhr88) [inline tools/plugins/com.liferay.ide.core/src/com/liferay/ide/core/LiferayCore.java]: no need this method

File: tools/plugins/com.liferay.ide.core/src/com/liferay/ide/core/event/LiferayResourceChangeManager.java [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+			}
+
+			return false;
+		}
+		else {
+			return resource instanceof IWorkspaceRoot;
+		}
+	}
+
+	private LiferayListenerRegistry _listenerRegistry() {
```
Comment (jtydhr88) [inline tools/plugins/com.liferay.ide.core/src/com/liferay/ide/core/event/LiferayResourceChangeManager.java]: no need this method

File: tools/plugins/com.liferay.ide.gradle.core/src/com/liferay/ide/gradle/core/LiferayGradleProject.java [hunk 7]
```diff
@@ -362,4 +393,7 @@ private IFolder _createResorcesFolder(IProject project) {
 
 	private static final String[] _IGNORE_PATHS = {".gradle", "build", "dist", "liferay-theme.json"};
 
+	private ImmutableSet<IPath> _importantResources;
```
Comment (jtydhr88) [inline tools/plugins/com.liferay.ide.gradle.core/src/com/liferay/ide/gradle/core/LiferayGradleProject.java]: why do we need this ImmutableSet or not common HashSet

File: tools/plugins/com.liferay.ide.gradle.core/src/com/liferay/ide/gradle/core/LiferayGradleProject.java [hunk 8]
```diff
@@ -59,10 +63,15 @@
  * @author Terry Jia
  * @author Andy Wu
  */
-public class LiferayGradleProject extends BaseLiferayProject implements IBundleProject, IResourceBundleProject {
+public class LiferayGradleProject
+	extends BaseLiferayProject implements IBundleProject, IResourceBundleProject, EventListener {
```
Comment (jtydhr88) [inline tools/plugins/com.liferay.ide.gradle.core/src/com/liferay/ide/gradle/core/LiferayGradleProject.java]: do we still need EventListener here if you already add it as the parent of ILiferayProject?

File: tools/plugins/com.liferay.ide.gradle.core/src/com/liferay/ide/gradle/core/LiferayGradleWorkspaceProject.java [hunk 9]
```diff
@@ -54,10 +58,16 @@
  * @author Simon Jiang
  * @author Terry Jia
  */
-public class LiferayGradleWorkspaceProject extends LiferayWorkspaceProject {
+public class LiferayGradleWorkspaceProject extends LiferayWorkspaceProject implements EventListener {
```
Comment (jtydhr88) [inline tools/plugins/com.liferay.ide.gradle.core/src/com/liferay/ide/gradle/core/LiferayGradleWorkspaceProject.java]: do we still need EventListener here if you already add it as the parent of ILiferayProject?

File: tools/plugins/com.liferay.ide.gradle.core/src/com/liferay/ide/gradle/core/LiferayGradleWorkspaceProject.java [hunk 10]
```diff
@@ -254,4 +286,7 @@ public void done(IJobChangeEvent event) {
 
 	private static final Set<IProject> _watchingProjects = new HashSet<>();
 
+	private ImmutableSet<IPath> _importantResources;
```
Comment (jtydhr88) [inline tools/plugins/com.liferay.ide.gradle.core/src/com/liferay/ide/gradle/core/LiferayGradleWorkspaceProject.java]: why do we need this ImmutableSet or not common HashSet

File: tools/plugins/com.liferay.ide.core/src/com/liferay/ide/core/IResourceBundleProject.java (hunk 2 above)
Comment (jtydhr88) [reply tools/plugins/com.liferay.ide.core/src/com/liferay/ide/core/IResourceBundleProject.java]: got it

**Comment by jtydhr88** [comment]
see comments and resend, thanks.

File: tools/plugins/com.liferay.ide.core/src/com/liferay/ide/core/LiferayCore.java (hunk 3 above)
Comment (Camork) [reply tools/plugins/com.liferay.ide.core/src/com/liferay/ide/core/LiferayCore.java]: Greg did.

**Pull request closed without merging**